### ISSUE TITLE
Replace Kafka write aggregation logic

### ISSUE THREAD

**Issue by domodwyer** [influxdata/influxdb_iox#5323]
# TL;DR

The aggregation logic in the write buffer is largely ineffective, costing us performance and likely contributing to the slow ingester recovery time (https://github.com/influxdata/conductor/issues/971) and high latency of kafka writes after rollouts (https://github.com/influxdata/conductor/issues/987). The complexity of this layer has also led hard-to-track-down bugs ([1][partition-bug], [2][flushing]).

We should replace or remove the aggregation logic.

[partition-bug]: https://github.com/influxdata/influxdb_iox/pull/4878
[flushing]: https://github.com/influxdata/influxdb_iox/pull/5113

## Analysis

The write buffer contains an [aggregation layer] that attempts to coalesce multiple DML writes together into one big DML write, before pushing the resulting batch of merged ops to kafka. 

The intent of this aggregation logic to increase the size of each DML write op sent over the wire, thereby increasing throughput (fewer messages, containing more rows) and to increase the effectiveness of compression (dictionary encoding in the `MutableBuffer`, block compression of payloads, etc).

Kafka/IOx is configured to allow a max message size of 10MiB however on average, the messages in Kafka have small payloads - the average uncompressed payload size is ~2103 bytes. We have data per kafka partition, and while there is some variation across partitions, almost all partitions (>75%) have an average message size of **just 2.3KiB**:

![image](https://user-images.githubusercontent.com/9275968/182859422-ae5bda19-f704-4589-952f-48ac28981e35.png)

<details><summary>
Quantiles of partition avg. payload size
</summary>

|     |        x|
|:----|--------:|
|0%   |   815.00|
|25%  |  1278.50|
|50%  |  1606.00|
|75%  |  2318.75|
|100% | 10845.00|
</details>

<details><summary>
Raw partition avg. payload size data
</summary>

| kafka_partition|avg_payload_size |
|---------------:|:----------------|
|               0|988              |
|               1|2.97KiB          |
|              10|7.73KiB          |
|              11|1.72KiB          |
|              12|2.56KiB          |
|              13|2.86KiB          |
|              14|1.62KiB          |
|              15|1.07KiB          |
|              16|1.24KiB          |
|              17|2.24KiB          |
|              18|1.99KiB          |
|              19|2.49KiB          |
|               2|2.04KiB          |
|              20|1.54KiB          |
|              21|1.22KiB          |
|              22|2.11KiB          |
|              23|1005             |
|              24|1.56KiB          |
|               3|1.38KiB          |
|               4|1.85KiB          |
|               5|1.61KiB          |
|               6|4.38KiB          |
|               7|1.36KiB          |
|               8|1.43KiB          |
|               9|1.15KiB          |
|              50|2.39KiB          |
|              51|1.35KiB          |
|              52|1.97KiB          |
|              53|2.75KiB          |
|              54|1.24KiB          |
|              55|1.15KiB          |
|              56|1.39KiB          |
|              57|1.25KiB          |
|              58|1.34KiB          |
|              59|1.85KiB          |
|              60|2.11KiB          |
|              61|1.24KiB          |
|              62|3.11KiB          |
|              63|1.70KiB          |
|              64|7.92KiB          |
|              65|1.98KiB          |
|              66|1.47KiB          |
|              67|1.24KiB          |
|              68|1.46KiB          |
|              69|1.41KiB          |
|              70|1.37KiB          |
|              71|10.59KiB         |
|              72|1.83KiB          |
|              73|1.59KiB          |
|              74|1.16KiB          |
|              25|5.68KiB          |
|              26|815              |
|              27|1.81KiB          |
|              28|2.36KiB          |
|              29|1.19KiB          |
|              30|1.40KiB          |
|              31|2.98KiB          |
|              32|954              |
|              33|1.17KiB          |
|              34|2.35KiB          |
|              35|1.30KiB          |
|              36|1.80KiB          |
|              37|2.76KiB          |
|              38|4.27KiB          |
|              39|1.33KiB          |
|              40|1.53KiB          |
|              41|1.38KiB          |
|              42|850              |
|              43|4.04KiB          |
|              44|1.52KiB          |
|              45|3.09KiB          |
|              46|2.03KiB          |
|              47|2.02KiB          |
|              48|1.33KiB          |
|              49|3.34KiB          |
|              75|1.52KiB          |
|              76|1.08KiB          |
|              77|1.12KiB          |
|              78|1.49KiB          |
|              79|1.33KiB          |
|              80|2.06KiB          |
|              81|2.37KiB          |
|              82|1.16KiB          |
|              83|960              |
|              84|925              |
|              85|1.12KiB          |
|              86|2.87KiB          |
|              87|1.57KiB          |
|              88|1.19KiB          |
|              89|3.10KiB          |
|              90|1.43KiB          |
|              91|2.03KiB          |
|              92|7.53KiB          |
|              93|1.35KiB          |
|              94|1.82KiB          |
|              95|2.64KiB          |
|              96|1.74KiB          |
|              97|991              |
|              98|924              |
|              99|1.79KiB          |
</details>

Given the small message sizes (and as part of a wider investigation into our Kafka perf) I [instrumented][metric-pr] the aggregation layer to record the number of DML ops merged into each kafka message:

![Screenshot 2022-08-05 at 13 39 02](https://user-images.githubusercontent.com/9275968/183070304-e7806c85-807f-4b0f-a7c8-a1bc17164066.png)

This shows that almost all DML writes pushed to Kafka are single DML writes, having never been merged with any other write before pushing the op to Kafka - for the common case, the aggregation logic provides no benefit but adds overhead / latency.

[metric-pr]:https://github.com/influxdata/influxdb_iox/pull/5307

[aggregation layer]:https://github.com/influxdata/influxdb_iox/blob/fc1870ff766cf35b8320aeb2bbfb41c093966546/write_buffer/src/kafka/aggregator.rs

## Why

The aggregation layer can only merge DML writes if they are for the same namespace, schema & have the same partition key:

https://github.com/influxdata/influxdb_iox/blob/fc1870ff766cf35b8320aeb2bbfb41c093966546/write_buffer/src/kafka/aggregator.rs#L114-L138

This is because the DML write contains both the namespace & partition key within. The aggregator waits ~5ms for other DML writes that can be merged before pushing the batch to Kafka.

The data gathered suggests our write workload doesn't satisfy the constraints of the current aggregation logic. This makes sense when you consider most users send a batch of writes to a bucket infrequently, rather than frequent small writes for the same bucket within the 5ms aggregation window.

## Proposal

Instead of merging together DML writes with the above constraints to produce a single DML write, change the aggregation logic (and wire format) to batch together multiple independent DML writes (i.e. `Vec<DmlWrite>`) per kafka partition.

### Pros

* Removes the namespace / schema / partition batching restriction - all write ops for a given partition can be batched leading to much larger messages
  * Increased throughput
  * Less aggregation overhead per dispatched message
* Preserves the atomicity of the DML op - less likely to introduce [bugs][partition-bug]
* Less messages hopefully leads to faster ingester recovery (https://github.com/influxdata/conductor/issues/971)
* Code [already exists][simple-batch] to test the cost of removing the DML merging aggregator in isolation 🎉

### Cons

* Requires changing the Kafka wire format
  * Blocked by https://github.com/influxdata/influxdb_iox/issues/4866 which is already necessary for https://github.com/influxdata/influxdb_iox/issues/4880 and https://github.com/influxdata/influxdb_iox/issues/4897

[simple-batch]: https://docs.rs/rskafka/latest/rskafka/client/producer/aggregator/struct.RecordAggregator.html

**Comment by tustvold** [comment]
> Code [already exists](https://docs.rs/rskafka/latest/rskafka/client/producer/aggregator/struct.RecordAggregator.html) tada

TBC this is not a message aggregator, but a `Record` aggregator. As such it is doing a subset of what the IOx aggregator is already doing. To clarify a single protobuf `DMLWrite` is written to a `Record`, and a group of `Vec<Record>` are then flushed to Kafka in one go, where each record is a separate "message".

As such switching to this aggregator is unlikely to yield any performance improvements, although I agree it may simplify the codebase

**Comment by alamb** [comment]
Question 1: I wonder if there is any value to increasing the buffer time from 5ms to something larger to see if there is any affect on the effectiveness of the aggregator?

**Comment by alamb** [comment]
Observation: Since there is already a non trivial amount of prefetching in the kafka protocol, I wonder if we could run a cheap experiment by removing all iox specific batching / aggregating and just sending each DML write as its own kafka message.

I think this is what @tustvold  is saying will happen at the kafka level even if we use https://docs.rs/rskafka/latest/rskafka/client/producer/aggregator/struct.RecordAggregator.html, 

My point is maybe we can start simple and then re-add aggregation logic if we found the overhead of multiple messages is too large. Also, if we just removed the DMLAggregator would the kafka format change?

**Comment by tustvold** [comment]
> My point is maybe we can start simple and then re-add aggregation logic if we found the overhead of multiple messages is too large.

I think it is worth a shot, I would not expect any material difference. The reasoning behind the `DMLAggregator` is it improves the throughput of Kafka by orders of magnitude if you push larger Records, as such the fact the aggregator is not working may be contributing to our Kafka performance issues. We can remove it, but part of me feels we should actually fix it instead...

> Also, if we just removed the DMLAggregator would the kafka format change?

No

**Comment by domodwyer** [comment]
> improves the throughput of Kafka by orders of magnitude if you push larger Records

This isn't happening with the current aggregator though, because the conditions in which we can coalesce DML ops and produce a large `Record` are so unlikely that it's never used in the common case.

I am suggesting we batch up a collection of DML writes (`Vec<DML>`) into a single `Record` and publish that to increase the size of the `Record` we send - do you think this will be problematic / unhelpful?

**Comment by domodwyer** [comment]
Spoke on slack - clarification:

> Code [already exists](https://docs.rs/rskafka/latest/rskafka/client/producer/aggregator/struct.RecordAggregator.html) 🎉

This will be to assess how much we loose by removing the DML merging aggregator without having to change any protocols / write any code and validate the numbers / analysis.

> batch together multiple independent DML writes (i.e. Vec<DmlWrite>)

Then this will be implemented to measure how effective this new aggregation strategy (potentially) is 👍

**Comment by domodwyer** [comment]
I've been working on https://github.com/influxdata/conductor/issues/987 to reduce the amount of batched write drops during a deployment, caused by high write latency to Kafka. The drops ticket also ties into this ticket, as the aggregation layer described in this ticket is adding overhead to the Kafka write, which in turn increase the write latency, and therefore increase the number of write drops and [external P95 latency](https://github.com/influxdata/conductor/issues/987#issuecomment-1219300228).

## Analysis

We currently measure two latencies that relate to Kafka:

1. **Pre-aggregation latency**

This latency measurement includes the overhead of the aggregation logic, plus all the latency of (2).

2. **Post-aggregation latency**

This measurement includes rskafka client overhead, serialisation, protocol overhead and network I/O.

```
                    ┌──────────────────┐
                    │   Router Logic   │
                    └──────────────────┘
                              │
                              │
                              ◎━━━━━━━━━━━━━━━━━━━ (1) Pre Aggregation
                              │                          Latency
                              │
             ╔  WriteBuffer  ═│═ ═ ═ ═ ═ ═ ═ ═
                              ▼               ║
             ║      ┌──────────────────┐
                    │    Aggregator    │      ║
             ║      └──────────────────┘
                              │               ║
             ║                ◎━━━━━━━━━━━━━━━━━━━ (2) Post Aggregation
                              │               ║          Latency
             ║                ▼
                    ┌──────────────────┐      ║
             ║      │   Kafka Client   │
                    └──────────────────┘      ║
             ║                │
                              │               ║
             ╚ ═ ═ ═ ═ ═ ═ ═ ═│═ ═ ═ ═ ═ ═ ═ ═
                              │
                              │
                              ▼
         
                           Kafka

```

This allows us to measure the overhead of the `Aggregator` step this ticket relates to.

Below is the P95 latency of the client `(2)` plotted alongside the latency of the aggregation + client `(1)`:

![image](https://user-images.githubusercontent.com/9275968/185373615-e7d663ec-27e4-402a-ab19-ff042b2f6b6c.png)

The difference in P95 latency before, and after the aggregator can be used to measure the overhead of the current aggregation logic (P95 latency diff):

![image](https://user-images.githubusercontent.com/9275968/185373689-e7a84213-683d-4916-bbcc-ad027c243875.png)

During deployments, the aggregation layer **adds 9 seconds** to the P95 latency, therefore improving this aggregation overhead can be a great benefit to the router / external P95 latency too.

**Comment by domodwyer** [comment]
# TL;DR

Experiment success! Replacing the old aggregator with a simpler implementation (https://github.com/influxdata/influxdb_iox/pull/5427) showed no change in average payload size, nor any effect on the HTTP LP write latency, confirming the old aggregator was not helping at all.

## Analysis

This confirms the DML-merging aggregator was ineffective with our workload, and that a far less complex aggregator performs identically.

### Average Payload Size

The distribution of average payload size per Kafka partition, across all partitions, is effectively the same[^1] (deployment spikes removed to compare steady state):

![plot](https://user-images.githubusercontent.com/9275968/186124450-f7a40985-f09e-45c2-afc4-7013054f4640.png)

### HTTP Write Latency (P95)

The P95 write endpoint latency also remained unchanged[^2]:

![plot](https://user-images.githubusercontent.com/9275968/186125484-831bff7d-c017-4b52-8753-2b29dc1f67df.png)

### Next Improvement

There have been no deployments since #5427 itself was deployed, but during the deployment of that PR, the routers saw similar spikes in the aggregation overhead (at P95):

![plot](https://user-images.githubusercontent.com/9275968/186128646-107f8edd-b6a6-421e-ad26-c191e00a7391.png)

The deployments happened at different times of day, but clearly the aggregation overhead remains. Given the simplistic implementation of actual aggregator implementation, this has lead me to investigate rskafka's usage of the aggregator (which happens before the "post aggregation latency" metric - meaning it is included in these overhead measurements.)

Though I have a suspicion as to the culprit, I will wait for more deployments to gather data before continuing into rskafka improvements. For now, I will continue working on removing blockers for the proposed `Vec<DmlWrite>` aggregator to increase Kafka payload size.


[^1]: Mean payload size increased by 132 bytes (p=4.13e-55) likely due to typical workload changes throughout the day
[^2]: Mean latency increased by 1.2ms (p=9.86e-60), an insignificant amount when investigating 10s of P95 latency

**Comment by domodwyer** [comment]
After https://github.com/influxdata/rskafka/pull/173 aggregation overhead was most recently measured as 121ms at P95, a significant reduction from ~9000ms previously!

Message sizes are relatively unchanged after the PR (not surprising), so increasing the batch sizes by changing the batching strategy remains an open task.
